// apollo-server-core/dist/plugin/schemaReporting/reportingOperationTypes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReportServerInfoErrorCode = void 0;
var ReportServerInfoErrorCode;
(function (ReportServerInfoErrorCode) {
    ReportServerInfoErrorCode["BOOT_ID_IS_NOT_VALID_UUID"] = "BOOT_ID_IS_NOT_VALID_UUID";
    ReportServerInfoErrorCode["BOOT_ID_IS_REQUIRED"] = "BOOT_ID_IS_REQUIRED";
    ReportServerInfoErrorCode["EXECUTABLE_SCHEMA_ID_IS_NOT_SCHEMA_SHA256"] = "EXECUTABLE_SCHEMA_ID_IS_NOT_SCHEMA_SHA256";
    ReportServerInfoErrorCode["EXECUTABLE_SCHEMA_ID_IS_REQUIRED"] = "EXECUTABLE_SCHEMA_ID_IS_REQUIRED";
    ReportServerInfoErrorCode["EXECUTABLE_SCHEMA_ID_IS_TOO_LONG"] = "EXECUTABLE_SCHEMA_ID_IS_TOO_LONG";
    ReportServerInfoErrorCode["GRAPH_VARIANT_DOES_NOT_MATCH_REGEX"] = "GRAPH_VARIANT_DOES_NOT_MATCH_REGEX";
    ReportServerInfoErrorCode["GRAPH_VARIANT_IS_REQUIRED"] = "GRAPH_VARIANT_IS_REQUIRED";
    ReportServerInfoErrorCode["LIBRARY_VERSION_IS_TOO_LONG"] = "LIBRARY_VERSION_IS_TOO_LONG";
    ReportServerInfoErrorCode["PLATFORM_IS_TOO_LONG"] = "PLATFORM_IS_TOO_LONG";
    ReportServerInfoErrorCode["RUNTIME_VERSION_IS_TOO_LONG"] = "RUNTIME_VERSION_IS_TOO_LONG";
    ReportServerInfoErrorCode["SERVER_ID_IS_TOO_LONG"] = "SERVER_ID_IS_TOO_LONG";
    ReportServerInfoErrorCode["USER_VERSION_IS_TOO_LONG"] = "USER_VERSION_IS_TOO_LONG";
})(ReportServerInfoErrorCode = exports.ReportServerInfoErrorCode || (exports.ReportServerInfoErrorCode = {}));
